### ISSUE TITLE
remove @ptrCast. add @elemCast

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#1106]
Right now, `@ptrCast` can change the pointer element, but it gives a compile error if you try to change the alignment or const-ness. We already have `@alignCast` to change the alignment only. Really, this cast is only meant to change the underlying element. So let's make that explicit by having the built-in function only accept the child element type as a parameter, and then it's obvious that the other pointer attributes are unmodified.

**Comment by alexnask** [comment]
I will work on this one if you don't mind.

**Comment by Hejsil** [comment]
How would we handle:
* `*T` -> `[*]T` -> `*T` 
* `*T` -> `fn(...) ...` -> `*T`
* `fn(...) ...` -> `fn(...) ...`

**Comment by alexnask** [comment]
`*T -> [*]T` currently does not work even with explicit casting, `*T -> *[1]T -> [*]T`.  
The rest do not explicitly cast either.  

I'm leaving this up here and will update with further work, will close if the proposal is significantly changed/not approved.

**Comment by alexnask** [comment]
@Hejsil  

Right, I was saying those cases do not work without @ptrCast, since you can't "just" cast from one of those types to another, I'm not disagreeing :)

**Comment by andrewrk** [comment]
> How would we handle:

* `*T` -> `[*]T` -> `*T` 

To get from `*T` to `[*]T` I propose a userland solution, that does `*T` -> `*[1]T` -> `[*]T` (these are all implicit casts)

To get from `[*]T` to `*T`, do this: `&foo[0]`

* `*T` -> `fn(...) ...` -> `*T`
* `fn(...) ...` -> `fn(...) ...`

maybe `@fnCast`? but it wouldn't do `fn(...) ...` -> `*T`.

**Comment by Hejsil** [comment]
I guess that could be `@intToPtr(*T, @ptrToInt(func))`. I think `fn` -> `*T` is really rare, while `*T` -> `fn` is more "common".

**Comment by Hejsil** [comment]
Discarding attributes is a very unsafe operations. You can easily discard attributes you don't want to discard with `@ptrCast` when you refactor code. It's to powerful. That is probably why `@ptrCast` actually can't discard all pointer attributes (like `const`).

`@elemCast` (or just change `@ptrCast`s behavior but not it's name) is a safer more controlled function. When you really want to do your unsafe operations, you always have the `@ptrToInt` `@intToPtr` pair. They can do everything.

**Comment by andrewrk** [comment]
Re-opening for reconsideration in light of two new developments:
 * The destination type parameter is eliminated (#5909)
 * Function pointers now use uniform syntax with pointers

**Comment by kocsis1david** [comment]
> When you really want to do your unsafe operations, you always have the `@ptrToInt` `@intToPtr` pair. They can do everything.

It would be nice if you didn't need two casts for that.

**Comment by andrewrk** [comment]
In modern Zig, `@ptrCast` can change three things:
* The element type
* The size (e.g. single/multi)
* The length of the slice (keeping byte quantity the same)

This proposal is no longer needed. https://github.com/ziglang/zig/issues/23935 is accepted to address the last bullet point.
